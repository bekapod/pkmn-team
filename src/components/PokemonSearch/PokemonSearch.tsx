import {
  ChangeEvent,
  KeyboardEvent,
  FunctionComponent,
  useCallback,
  useEffect,
  useRef,
  useState,
  useMemo
} from 'react';
import { FixedSizeList as List, ListChildComponentProps } from 'react-window';
import type { CombinedError } from 'urql';
import { Pokemon } from '~/generated/graphql';
import { Autocomplete, AutocompleteDropdown } from '../Autocomplete';
import { CenteredRow } from '../CenteredRow';
import { ErrorMessage } from '../ErrorMessage';
import { GiantInput } from '../GiantInput';
import { LoadingIcon } from '../LoadingIcon';
import { PokemonLine } from '../PokemonLine';

export type PokemonSearchProps = {
  setCurrentSearchPokemon: (pokemon: Pokemon) => void;
  pokemon: Pokemon[];
  isLoading?: boolean;
  error?: CombinedError;
};

type RowProps = ListChildComponentProps & {
  data: {
    pokemon: Pokemon;
    isHighlighted: boolean;
    onResultClick: (pkmn: Pokemon, index: number) => void;
  }[];
};

const Row = ({ index, data, style, isScrolling, ...itemProps }: RowProps) => {
  const { pokemon, isHighlighted, onResultClick: onClick } = data[index];
  return (
    <PokemonLine
      aria-selected={isHighlighted}
      data-testid={`autocomplete-result-${pokemon.pokedex_id}`}
      pokemon={pokemon}
      {...itemProps}
      onClick={() => onClick(pokemon, index)}
      style={{
        ...style,
        backgroundColor: isHighlighted
          ? 'var(--colors-yellow-vivid-100)'
          : 'initial',
        top: `calc(${style?.top}px + var(--spacing-1))`
      }}
    />
  );
};

export const PokemonSearch: FunctionComponent<PokemonSearchProps> = ({
  setCurrentSearchPokemon,
  pokemon,
  isLoading,
  error
}) => {
  const listContainer = useRef<List>(null);
  const [highlightedIndex, setHighlightedIndex] = useState(0);
  const [inputValue, setInputValue] = useState('');
  const [filteredList, setFilteredList] = useState(pokemon);
  const itemHeight = 96;

  useEffect(() => {
    listContainer.current?.scrollToItem(highlightedIndex, 'center');
  }, [highlightedIndex]);

  const onResultClick = useCallback(
    (pkmn: Pokemon, index: number) => {
      setCurrentSearchPokemon(pkmn);
      setHighlightedIndex(index);
    },
    [setCurrentSearchPokemon]
  );

  const updateList = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const { value } = e.currentTarget;
      setFilteredList(pokemon.filter(({ name }) => name.includes(value)));
      setHighlightedIndex(0);
      setInputValue(value);
    },
    [pokemon]
  );

  const keyboardNavigation = useCallback(
    (e: KeyboardEvent<HTMLInputElement>) => {
      if (e.key === 'Enter') {
        const selectedPokemon = filteredList[highlightedIndex];

        if (selectedPokemon) {
          setCurrentSearchPokemon(selectedPokemon);
        }

        return;
      }

      if (['ArrowUp', 'ArrowDown'].includes(e.key)) {
        if (e.key === 'ArrowUp') {
          setHighlightedIndex(val => {
            if (val !== 0) return val - 1;
            return val;
          });
        }

        if (e.key === 'ArrowDown') {
          setHighlightedIndex(val => {
            if (val < filteredList.length) return val + 1;
            return val;
          });
        }
      }
    },
    [filteredList, highlightedIndex, setCurrentSearchPokemon]
  );

  const itemData = useMemo(
    () =>
      filteredList.map((pokemon, idx) => ({
        pokemon,
        isHighlighted: idx === highlightedIndex,
        onResultClick
      })),
    [filteredList, highlightedIndex, onResultClick]
  );

  if (isLoading) {
    return (
      <CenteredRow stackVertically>
        <LoadingIcon isSpinner />
      </CenteredRow>
    );
  }

  if (!isLoading && error) {
    return (
      <CenteredRow stackVertically>
        <ErrorMessage>{error.message}</ErrorMessage>
      </CenteredRow>
    );
  }

  return (
    <Autocomplete>
      <GiantInput
        aria-label="Find Pokemon by name"
        placeholder="Find by name"
        isFullWidth
        value={inputValue}
        onKeyDown={keyboardNavigation}
        onChange={updateList}
      />
      <AutocompleteDropdown>
        <List
          ref={listContainer}
          height={itemHeight * 5}
          itemSize={itemHeight}
          itemCount={itemData.length}
          itemData={itemData}
          width={500}
          className="w-full!"
        >
          {Row}
        </List>
      </AutocompleteDropdown>
    </Autocomplete>
  );
};
