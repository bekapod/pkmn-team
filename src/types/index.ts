export type Type =
  | "NORMAL"
  | "FIGHTING"
  | "FLYING"
  | "POISON"
  | "GROUND"
  | "ROCK"
  | "BUG"
  | "GHOST"
  | "STEEL"
  | "FIRE"
  | "WATER"
  | "GRASS"
  | "ELECTRIC"
  | "PSYCHIC"
  | "ICE"
  | "DRAGON"
  | "DARK"
  | "FAIRY"
  | "UNKNOWN"
  | "SHADOW";

export interface IPokemon {
  id: string;
  pokedexId: number;
  name: string;
  types: Type[];
  sprite: string;
}

export interface ITeamMember {
  id: string;
  pokemon: IPokemon;
}

export interface ITeam {
  id: string;
  name: string;
  members: ITeamMember[];
  createdAt: string;
}

export interface ITeamBuilderState {
  name: string;
  members: { [key: string]: ITeamMember };
}

export interface IPokemonSearchState {
  currentSelection?: IPokemon;
  highlightedIndex: number;
  inputValue?: string;
  filteredList: IPokemon[];
  unfilteredList: IPokemon[];
}

export interface IState {
  teamBuilder: ITeamBuilderState;
  pokemonSearch: IPokemonSearchState;
}
