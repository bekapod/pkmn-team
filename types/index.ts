import {} from "styled-components/cssprop"; // eslint-disable-line import/no-unresolved

export type Type =
  | "NORMAL"
  | "FIGHTING"
  | "FLYING"
  | "POISON"
  | "GROUND"
  | "ROCK"
  | "BUG"
  | "GHOST"
  | "STEEL"
  | "FIRE"
  | "WATER"
  | "GRASS"
  | "ELECTRIC"
  | "PSYCHIC"
  | "ICE"
  | "DRAGON"
  | "DARK"
  | "FAIRY"
  | "UNKNOWN"
  | "SHADOW";

export interface Move {
  name: string;
  types: { name: Type }[];
}

export interface PokemonMove {
  levelLearnedAt: number;
  learnMethod: string;
  version: string;
  move: Move;
}

export interface MoveVariation {
  levelLearnedAt: number;
  learnMethod: string;
  version: string;
}

export interface DeduplicatedMove {
  name: string;
  types: { name: Type }[];
  variations: MoveVariation[];
}

export interface Pokemon {
  id: string;
  pokedexId: number;
  name: string;
  types: { name: Type }[];
  sprite: string;
  moves: PokemonMove[];
}

export interface TeamMember {
  id: string;
  order: number;
  pokemon: Pokemon;
}

export interface Team {
  id: string;
  name: string;
  members: TeamMember[];
  insertedAt: string;
}

export interface TeamBuilderState {
  name: string;
  members: { [key: string]: TeamMember };
}

export interface PokemonSearchState {
  currentSelection?: Pokemon;
  highlightedIndex: number;
  inputValue?: string;
  filteredList: Pokemon[];
  unfilteredList: Pokemon[];
}

export interface State {
  teamBuilder: TeamBuilderState;
  pokemonSearch: PokemonSearchState;
}
