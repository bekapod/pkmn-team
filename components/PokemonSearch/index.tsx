import { ApolloError } from "apollo-client";
import { size } from "lodash/fp";
import React, { PureComponent } from "react";
import { FixedSizeList, FixedSizeList as List } from "react-window";
import { css } from "styled-components/macro";
import * as variables from "../../helpers/variables";
import { IPokemon } from "../../types";
import Autocomplete, { AutocompleteDropdown } from "../Autocomplete";
import CenteredRow from "../CenteredRow";
import ErrorMessage from "../ErrorMessage";
import GiantInput from "../GiantInput";
import LoadingIcon from "../LoadingIcon";
import PokemonLine from "../PokemonLine";

interface IProps {
  pokemon: IPokemon[];
  highlightedIndex: number;
  inputValue?: string;
  filteredList: IPokemon[];
  loading?: boolean;
  error?: ApolloError;
  setCurrentSelection: (pokemon: IPokemon) => void;
  setHighlightedIndex: (index: number) => void;
  setInputValue: (value: string) => void;
  setUnfilteredList: (pokemon: IPokemon[]) => void;
}

const itemHeight = variables.spacing.xxl;

const resultItem = (
  pokemon: IPokemon[],
  highlightedIndex: number,
  onClick: any
) => ({ index, style, ...itemProps }: { index: number; style: any }) => {
  const pkmn = pokemon[index];
  const isHighlighted = index === highlightedIndex;
  const props = {
    ...itemProps,
    onClick: () => onClick(pkmn),
    style: {
      ...style,
      backgroundColor: isHighlighted ? variables.colors.grayLight : "initial",
      top: `calc(${style.top}px + ${variables.spacing.xs}px)`
    }
  };

  return (
    <PokemonLine
      aria-selected={index === highlightedIndex}
      data-testid={`autocomplete-result-${pkmn.pokedexId}`}
      pokemon={pkmn}
      {...props}
    />
  );
};

class PokemonSearch extends PureComponent<IProps> {
  public listContainer: React.RefObject<FixedSizeList> = React.createRef();

  constructor(props: IProps) {
    super(props);

    this.keyboardNavigation = this.keyboardNavigation.bind(this);
    this.setInputValue = this.setInputValue.bind(this);
  }

  public componentWillMount() {
    if (!this.props.loading) {
      this.props.setUnfilteredList(this.props.pokemon);
      this.props.setInputValue("");
    }
  }

  public componentDidMount() {
    if (this.listContainer.current) {
      this.listContainer.current.scrollToItem(
        this.props.highlightedIndex,
        "center"
      );
    }
  }

  public componentDidUpdate(prevProps: IProps) {
    if (prevProps.loading && !this.props.loading) {
      this.props.setUnfilteredList(this.props.pokemon);
      this.props.setInputValue("");
    }

    if (this.listContainer.current) {
      this.listContainer.current.scrollToItem(
        this.props.highlightedIndex,
        "center"
      );
    }
  }

  public keyboardNavigation(e: React.KeyboardEvent) {
    const { highlightedIndex } = this.props;

    if (e.key === "Enter") {
      this.props.setCurrentSelection(this.props.filteredList[highlightedIndex]);
      return;
    }

    let newHighlightedIndex = 0;

    if (["ArrowUp", "ArrowDown"].includes(e.key)) {
      if (e.key === "ArrowUp") {
        if (highlightedIndex !== 0) {
          newHighlightedIndex = highlightedIndex - 1;
        }
      }

      if (e.key === "ArrowDown") {
        if (highlightedIndex < size(this.props.filteredList)) {
          newHighlightedIndex = highlightedIndex + 1;
        }
      }

      this.props.setHighlightedIndex(newHighlightedIndex);
    }
  }

  public setInputValue(e: React.ChangeEvent<HTMLInputElement>) {
    const { value } = e.target;

    this.props.setHighlightedIndex(0);
    this.props.setInputValue(value);
  }

  public render() {
    if (this.props.loading) {
      return (
        <CenteredRow stackVertically={true}>
          <LoadingIcon spinner={true} />
        </CenteredRow>
      );
    }

    if (!this.props.loading && this.props.error) {
      return (
        <CenteredRow stackVertically={true}>
          <ErrorMessage>{this.props.error.message}</ErrorMessage>
        </CenteredRow>
      );
    }

    return (
      <Autocomplete>
        <GiantInput
          arial-label="Find Pokemon by name"
          placeholder="Find by name"
          css={css`
            max-width: none;
          `}
          value={this.props.inputValue || ""}
          onKeyDown={this.keyboardNavigation}
          onChange={this.setInputValue}
        />
        <AutocompleteDropdown>
          <List
            ref={this.listContainer}
            height={itemHeight * 5}
            itemSize={itemHeight}
            itemCount={size(this.props.filteredList)}
            width={500}
          >
            {resultItem(
              this.props.filteredList,
              this.props.highlightedIndex,
              this.props.setCurrentSelection
            )}
          </List>
        </AutocompleteDropdown>
      </Autocomplete>
    );
  }
}

export default PokemonSearch;
