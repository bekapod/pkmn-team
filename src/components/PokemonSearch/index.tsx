import Downshift from "downshift";
import { toLower } from "lodash/fp";
import React from "react";
import { capitalizePokemonName } from "../../helpers/general";
import * as variables from "../../helpers/variables";
import { IPokemon } from "../../types";
import Autocomplete, { AutocompleteDropdown } from "../Autocomplete";
import GiantInput from "../GiantInput";
import PokemonLine from "../PokemonLine";

interface IProps {
  pokemon: IPokemon[];
  currentSearchPokemon?: IPokemon;
  setCurrentSearchPokemon: (pokemon: IPokemon) => void;
}

const PokemonSearch = ({
  pokemon,
  currentSearchPokemon,
  setCurrentSearchPokemon
}: IProps) => {
  const onChange = (selectedPkmn: IPokemon) =>
    setCurrentSearchPokemon(selectedPkmn);
  const itemToString = (pkmn: IPokemon) => capitalizePokemonName(pkmn);
  return (
    <div>
      <Downshift onChange={onChange} itemToString={itemToString}>
        {({
          getRootProps,
          getInputProps,
          getItemProps,
          isOpen,
          inputValue,
          highlightedIndex,
          selectedItem
        }) => (
          <Autocomplete {...getRootProps()}>
            <GiantInput
              {...getInputProps()}
              arial-label="Choose a Pokemon"
              placeholder="Choose a Pokemon"
            />
            {isOpen ? (
              <AutocompleteDropdown>
                {pokemon
                  .filter(
                    pkmn =>
                      !inputValue || pkmn.name.includes(toLower(inputValue))
                  )
                  .map((pkmn, index) => (
                    <div
                      {...getItemProps({
                        index,
                        item: pkmn,
                        key: pkmn.pokedexId,
                        style: {
                          backgroundColor:
                            highlightedIndex === index
                              ? variables.colors.grayLight
                              : variables.colors.white,
                          fontWeight: selectedItem === pkmn ? "bold" : "normal"
                        }
                      })}
                    >
                      <PokemonLine pokemon={pkmn} />
                    </div>
                  ))}
              </AutocompleteDropdown>
            ) : null}
          </Autocomplete>
        )}
      </Downshift>
    </div>
  );
};

export default PokemonSearch;
