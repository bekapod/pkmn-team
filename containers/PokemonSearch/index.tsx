import { getOr } from "lodash/fp";
import React from "react";
import { adopt } from "react-adopt";
import { Mutation, OperationVariables, Query, QueryResult } from "react-apollo";
import PokemonSearch from "../../components/PokemonSearch";
import { setCurrentSearchPokemon } from "../../mutations/search";
import { getPokemon } from "../../queries/pokemon";
import { Pokemon } from "../../types";

interface QueryProps {
  getAllPokemonQuery: QueryResult<{ pokemon: Pokemon[] }, OperationVariables>;
  setCurrentSearchPokemonMutation: ({
    variables: { pokemon }
  }: {
    variables: { pokemon: Pokemon };
  }) => void;
}

const queries = {
  getAllPokemonQuery: ({ render }: any) => (
    <Query query={getPokemon}>{render}</Query>
  )
};

const mutations = {
  setCurrentSearchPokemonMutation: ({ render }: any) => (
    <Mutation mutation={setCurrentSearchPokemon}>{render}</Mutation>
  )
};

const WithQueries = adopt({ ...queries, ...mutations });

const PokemonSearchContainer = (): JSX.Element => (
  <WithQueries>
    {({
      getAllPokemonQuery: { data, loading, error },
      setCurrentSearchPokemonMutation
    }: QueryProps) => {
      return (
        <PokemonSearch
          pokemon={getOr([], "pokemon", data)}
          setCurrentSearchPokemon={setCurrentSearchPokemonMutation}
          loading={loading}
          error={error}
        />
      );
    }}
  </WithQueries>
);

export default PokemonSearchContainer;
