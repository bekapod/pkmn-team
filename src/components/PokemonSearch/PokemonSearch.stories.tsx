import { Meta, Story } from '@storybook/react/types-6-0';
import { PokemonSearch, PokemonSearchProps } from './PokemonSearch';

export default {
  title: 'Components/PokemonSearch',
  component: PokemonSearch,
  argTypes: { setCurrentSearchPokemon: { action: 'setCurrentSearchPokemon' } }
} as Meta<PokemonSearchProps>;

export const pokemonSearch: Story<PokemonSearchProps> = args => (
  <PokemonSearch {...args} />
);
