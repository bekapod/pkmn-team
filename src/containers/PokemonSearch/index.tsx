import { getOr } from "lodash/fp";
import React, { PureComponent } from "react";
import { Query } from "react-apollo";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import * as pokemonSearchActions from "../../actions/pokemonSearch";
import PokemonSearch from "../../components/PokemonSearch";
import { getAllPokemon } from "../../queries/pokemon";
import { IPokemon } from "../../types";

interface IProps {
  setCurrentSelection: (_: IPokemon) => void;
}

class PokemonSearchContainer extends PureComponent<IProps> {
  public render() {
    return (
      <Query query={getAllPokemon}>
        {({ data, loading, error }) => {
          return (
            <PokemonSearch
              {...this.props}
              pokemon={getOr([], "allPokemon", data)}
              loading={loading}
              error={error}
            />
          );
        }}
      </Query>
    );
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(pokemonSearchActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PokemonSearchContainer);
