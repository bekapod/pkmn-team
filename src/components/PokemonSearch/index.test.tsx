import React from "react";
// tslint:disable-next-line:no-implicit-dependencies
import renderer from "react-test-renderer";
// tslint:disable-next-line:no-implicit-dependencies
import { cleanup, fireEvent, render } from "react-testing-library";
import PokemonSearch from ".";
import { IPokemon } from "../../types";

describe("<PokemonSearch />", () => {
  const pokemon: IPokemon[] = [
    {
      id: "4",
      name: "charmander",
      pokedexId: 4,
      sprite: "4.png",
      types: ["FIRE"]
    },
    {
      id: "25",
      name: "pikachu",
      pokedexId: 25,
      sprite: "25.png",
      types: ["ELECTRIC"]
    },
    {
      id: "93",
      name: "haunter",
      pokedexId: 93,
      sprite: "93.png",
      types: ["GHOST", "POISON"]
    }
  ];

  afterEach(() => {
    cleanup();
  });

  it("renders without crashing", () => {
    const setCurrentSearchPokemon = () => null;
    const tree = renderer.create(
      <PokemonSearch
        pokemon={pokemon}
        setCurrentSearchPokemon={setCurrentSearchPokemon}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("shows list of matching pokemon when user begins to search", () => {
    const setCurrentSearchPokemon = () => null;
    const { getByPlaceholderText, queryByText } = render(
      <PokemonSearch
        pokemon={pokemon}
        setCurrentSearchPokemon={setCurrentSearchPokemon}
      />
    );

    fireEvent.change(getByPlaceholderText(/Choose a Pokemon/), {
      target: { value: "hau" }
    });

    expect(queryByText(/Haunter/)).toBeTruthy();
    expect(queryByText(/Pikachu/)).toBeFalsy();
    expect(queryByText(/Charmander/)).toBeFalsy();
  });

  it("shows full list of pokemon when user has cleared their search", () => {
    const setCurrentSearchPokemon = () => null;
    const { getByPlaceholderText, queryByText } = render(
      <PokemonSearch
        pokemon={pokemon}
        setCurrentSearchPokemon={setCurrentSearchPokemon}
      />
    );

    fireEvent.change(getByPlaceholderText(/Choose a Pokemon/), {
      target: { value: "hau" }
    });

    fireEvent.change(getByPlaceholderText(/Choose a Pokemon/), {
      target: { value: "" }
    });

    expect(queryByText(/Haunter/)).toBeTruthy();
    expect(queryByText(/Pikachu/)).toBeTruthy();
    expect(queryByText(/Charmander/)).toBeTruthy();
  });

  it("doesn't show any pokemon when none are matched", () => {
    const setCurrentSearchPokemon = () => null;
    const { getByPlaceholderText, queryByText } = render(
      <PokemonSearch
        pokemon={pokemon}
        setCurrentSearchPokemon={setCurrentSearchPokemon}
      />
    );

    fireEvent.change(getByPlaceholderText(/Choose a Pokemon/), {
      target: { value: "blah" }
    });

    expect(queryByText(/Haunter/)).toBeFalsy();
    expect(queryByText(/Pikachu/)).toBeFalsy();
    expect(queryByText(/Charmander/)).toBeFalsy();
  });

  it("calls setCurrentSearchPokemon when user clicks a pokemon result", () => {
    const setCurrentSearchPokemon = jest.fn();
    const { getByPlaceholderText, getByTestId } = render(
      <PokemonSearch
        pokemon={pokemon}
        setCurrentSearchPokemon={setCurrentSearchPokemon}
      />
    );

    expect(setCurrentSearchPokemon).toBeCalledTimes(0);

    fireEvent.change(getByPlaceholderText(/Choose a Pokemon/), {
      target: { value: "hau" }
    });

    expect(setCurrentSearchPokemon).toBeCalledTimes(0);

    fireEvent.click(getByTestId("autocomplete-result-93"));

    expect(setCurrentSearchPokemon).toBeCalledTimes(1);
    expect(setCurrentSearchPokemon).toBeCalledWith(pokemon[2]);
  });
});
