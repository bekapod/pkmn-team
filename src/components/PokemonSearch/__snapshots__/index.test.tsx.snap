// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonSearch /> renders without crashing 1`] = `
<div>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    aria-owns={null}
    className="Autocomplete-sc-1wak8g9-1 cILwaC"
    role="combobox"
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-controls={null}
      aria-labelledby="downshift-0-label"
      autoComplete="off"
      className="TextInput-sc-149814h-0 GiantInput-sc-1mhmfj5-0 CEtIK"
      id="downshift-0-input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Choose a Pokemon"
      value=""
    />
  </div>
</div>
`;
